design/design-doc-composition-functions.md

```diff
@@ -209,570 +738,353 @@ metadata:
-
-This document proposes that containerized functions support Crossplane [package
-metadata][package-meta] in the form of a `package.yaml` file at the root of the
-flattened filesystem and/or the OCI layer annotated as `io.crossplane.xpkg:
-base` per the [xpkg spec][xpkg-spec]. This `package.yaml` file would contain a
-custom-resource-like YAML document of type `Function.meta.pkg.crossplane.io`.
-
-Unlike `Configuration` and `Provider` packages, `Function` packages would not
-actually be processed by the Crossplane package manager but rather by the
-Composition (`apiextensions`) machinery. In practice Crossplane would be
-ignorant of the `package.yaml` file; it would exist purely as a way to attach
-"package-like" metadata to containerized Crossplane functions. Therefore, unlike
-the existing package types the `package.yaml` would contain no `spec` section.
-
-An example `package.yaml` might look like:
+A key benefit of Functions is that composition logic is decoupled from the core
+Crossplane release cycle. Moving composition logic out-of-tree and versioning it
+separately from Crossplane allows us to iterate faster, and experiment more
+freely.
+
+P&T style Composition could enjoy these benefits if it were not a special case,
+and were just another Function. Imagine for example wanting a new type of patch
+or a new type of transform and being able to simply fork the implementation to
+experiment without affecting everyone who uses Crossplane.
+
+P&T Composition as a Function might look like this:
 
 ```yaml
-# Required. Must be as below.
-apiVersion: meta.pkg.crossplane.io/v1alpha1
-# Required. Must be as below.
-kind: Function
-# Required.
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
 metadata:
-  # Required. Must comply with Kubernetes API conventions.
-  name: function-example
-  # Optional. Must comply with Kubernetes API conventions.
-  annotations:
-    meta.crossplane.io/source: https://github.com/negz/example-fn
-    meta.crossplane.io/description: An example function
+  name: example
+spec:
+  compositeTypeRef:
+    apiVersion: database.example.org/v1alpha1
+    kind: AcmeCoDatabase
+  functions:
+  - name: patch-and-transform
+    functionRef:
+      name: patch-and-transform
+    config:
+      apiVersion: apiextensions.crossplane.io/v1
+      kind: Resources
+      resources:
+      - name: cloudsqlinstance
+        base:
+          apiVersion: database.gcp.crossplane.io/v1beta1
+          kind: CloudSQLInstance
+          spec:
+            forProvider:
+              databaseVersion: POSTGRES_9_6
+              region: us-central1
+              settings:
+                tier: db-custom-1-3840
+                dataDiskType: PD_SSD
+        patches:
+        - type: FromCompositeFieldPath
+          fromFieldPath: spec.parameters.storageGB
+          toFieldPath: spec.forProvider.settings.dataDiskSizeGb
 ```
 
+Removing P&T support from the Composition type would be a breaking API change -
```

Comment:
could they exist simultaneously? e.g. P&T section still exists, but could be left totally empty, while the Functions based P&T approach is also supported within the same Composition?

Reply:
Yeah, that's close to what I was suggesting in the following sentence.